pandas/core/dtypes/base.py

```diff
@@ -0,0 +1,119 @@
+"""Extend pandas with custom array types"""
+import abc
+
+from pandas.compat import add_metaclass
+
+
+@add_metaclass(abc.ABCMeta)
+class ExtensionDtype(object):
+    """A custom data type for your array.
+    """
+
+    def __str__(self):
+        return self.name
+
+    @property
+    @abc.abstractmethod
+    def type(self):
+        # type: () -> type
+        """The scalar type for your array, e.g. ``int``
+
+        It's expected ``ExtensionArray[item]`` returns an instance
+        of ``ExtensionDtype.type`` for scalar ``item``.
+        """
+
+    @property
+    def kind(self):
+        # type () -> str
+        """A character code (one of 'biufcmMOSUV'), default 'O'
+
+        This should match the NumPy dtype used when your array is
+        converted to an ndarray, which is probably 'O' for object if
+        your extension type cannot be represented as a built-in NumPy
+        type.
+
+        See Also
+        --------
+        numpy.dtype.kind
+        """
+        return 'O'
+
+    @property
+    @abc.abstractmethod
+    def name(self):
+        # type: () -> str
+        """A string identifying the data type.
+
+        Will be used for display in, e.g. ``Series.dtype``
+        """
+
+    @property
+    def names(self):
+        # type: () -> Optional[List[str]]
```

Comment:
what is this for?

Reply:
We use it in `DataFrame.__init__`

```python
        elif isinstance(data, (np.ndarray, Series, Index)):
           if data.dtype.names:
```

I could of course modify that, but perhaps as a followup. For now having a non-abstract names property seems OK.